File: its-my-training-day/leetcode-top-75/questions/bit-manipulation/bitwise/counting-bits/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const countingBits = (n: number): number[] => {
  const result = new Array(n + 1).fill(0);
  for (let i = 1; i <= n; i += 1) {
    result[i] = result[i >> 1] + (i & 1);
  }
  return result;
};

File: its-my-training-day/leetcode-top-75/questions/bit-manipulation/bitwise/counting-bits/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def counting_bits(n: int) -> List[int]:
    result = [0] * (n + 1)
    i = 1
    while i <= n:
        result[i] = result[i >> 1] + (i & 1)
        i += 1
    return result
